geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/RPushDUnitTest.java

```diff
@@ -0,0 +1,211 @@
+import static org.apache.geode.test.dunit.rules.RedisClusterStartupRule.BIND_ADDRESS;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.IntStream;
+
+import com.google.common.collect.Streams;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import redis.clients.jedis.HostAndPort;
+import redis.clients.jedis.JedisCluster;
+
+import org.apache.geode.internal.cache.BucketDump;
+import org.apache.geode.internal.cache.InternalCache;
+import org.apache.geode.internal.cache.PartitionedRegion;
+import org.apache.geode.redis.ConcurrentLoopingThreads;
+import org.apache.geode.redis.internal.services.RegionProvider;
+import org.apache.geode.test.dunit.rules.ClusterStartupRule;
+import org.apache.geode.test.dunit.rules.MemberVM;
+import org.apache.geode.test.dunit.rules.RedisClusterStartupRule;
+import org.apache.geode.test.junit.rules.ExecutorServiceRule;
+
+public class RPushDUnitTest {
+  @ClassRule
+  public static RedisClusterStartupRule clusterStartUp = new RedisClusterStartupRule();
+
+  @ClassRule
+  public static ExecutorServiceRule executor = new ExecutorServiceRule();
+
+  private static JedisCluster jedis;
+  private static int locatorPort;
+  private static int redisServerPort;
+  private static MemberVM server1;
+
+  @BeforeClass
+  public static void testSetup() {
+    MemberVM locator = clusterStartUp.startLocatorVM(0);
+    locatorPort = locator.getPort();
+    server1 = clusterStartUp.startRedisVM(1, locatorPort);
+    clusterStartUp.startRedisVM(2, locatorPort);
+    clusterStartUp.startRedisVM(3, locatorPort);
+
+    redisServerPort = clusterStartUp.getRedisPort(3);
+    jedis = new JedisCluster(new HostAndPort(BIND_ADDRESS, redisServerPort), 10_000);
+  }
+
+  @After
+  public void cleanup() {
+    clusterStartUp.flushAll();
+  }
+
+  @AfterClass
+  public static void teardown() {
+    jedis.close();
+  }
+
+  @Test
+  public void givenBucketsMovedDuringRPush_elementsAreAddedAtomically() throws Exception {
+    AtomicBoolean running = new AtomicBoolean(true);
+    String KEY = "key";
+
+    Future<?> movingFuture = executor.submit(() -> {
+      try {
+        for (int i = 0; i < 20; i++) {
+          clusterStartUp.moveBucketForKey(KEY);
+          Thread.sleep(500);
+        }
+      } finally {
+        running.set(false);
+      }
+    });
+
+    int i = 0;
+    while (running.get()) {
+      jedis.rpush(KEY, value(i++), value(i++), value(i++), value(i++), value(i++));
+    }
+    movingFuture.get();
+
+    compareBuckets();
+
+    String popped;
+    int j = 0;
+    while ((popped = jedis.lpop(KEY)) != null) {
+      assertThat(popped).isEqualTo(value(j));
+      j++;
+    }
+  }
+
+  @Test
+  public void concurrentRPush_behavesCorrectly() {
```

Comment:
I think this test could/should be an IntegrationTest, since no distributed behaviour is being tested.